google/cloud/storage/_helpers.py

```diff
@@ -312,3 +312,25 @@ def _convert_to_timestamp(value):
     utc_naive = value.replace(tzinfo=None) - value.utcoffset()
     mtime = (utc_naive - datetime(1970, 1, 1)).total_seconds()
     return mtime
+
+
+def _bucket_bound_hostname_url(host, scheme=None, end_slash=False):
+    """Helper to build bucket bound hostname URL.
+
+    :type host: str
+    :param host: Host name.
+
+    :type scheme: str
+    :param scheme: (Optional) Web scheme. If passed, use it
+                   as a scheme in the result URL.
+
+    :type end_slash: bool
+    :param end_slash: (Optional) If True, add "/" slash to
+                      the end of the result URL.
+    """
+    if ":" in host:
+        return host
+
+    return "{scheme}://{host}{slash}".format(
+        scheme=scheme, host=host, slash="/" if end_slash else ""
+    )
```

Comment:
In V4 POST policies we're using URLs with end slash (that's what expected in conformance tests), but in other places URLs doesn't add it (and if I'll add it, several tests will fail). Keeping this behavior in the helper with `end_slash` arg. As it's just an internal helper I think we can afford it.

Reply:
What is the difference between the failures in this case?

Reply:
If we'll add a slash permanently:
`Test_Bucket.test_generate_signed_url_v4_w_bucket_bound_hostname_w_bare_hostname()` fails on assert:
```
AssertionError: expected call not found.
E         Expected: generate_signed_url_v4(None, resource='/', expiration=datetime.datetime(2020, 5, 15, 9, 28, 17, 120499, tzinfo=<UTC>), api_access_endpoint='http://cdn.example.com', method='GET', headers=None, query_parameters=None)
E           Actual: generate_signed_url_v4(None, resource='/', expiration=datetime.datetime(2020, 5, 15, 9, 28, 17, 120499, tzinfo=<UTC>), api_access_endpoint='http://cdn.example.com/', method='GET', headers=None, query_parameters=None)
```
`Test_Blob.test_generate_signed_url_v4_w_bucket_bound_hostname_w_bare_hostname()` fails on:
```
AssertionError: expected call not found.
E         Expected: generate_signed_url_v4(<object object at 0x000001515522AF50>, resource='/blob-name', expiration=datetime.datetime(2020, 5, 15, 9, 28, 15, 798333, tzinfo=<UTC>), api_access_endpoint='http://cdn.example.com', method='GET', content_md5=None, content_type=None, response_type=None, response_disposition=None, generation=None, headers=None, query_parameters=None, access_token=None, service_account_email=None)
E           Actual: generate_signed_url_v4(<object object at 0x000001515522AF50>, resource='/blob-name', expiration=datetime.datetime(2020, 5, 15, 9, 28, 15, 798333, tzinfo=<UTC>), api_access_endpoint='http://cdn.example.com/', method='GET', content_md5=None, content_type=None, response_type=None, response_disposition=None, generation=None, headers=None, query_parameters=None, service_account_email=None, access_token=None)
```
So, the `api_access_endpoint` doesn't match because of the end slash.

-

If we'll delete the end slash permanently, than `test_conformance_post_policy` fails with `POST Policy Simple Bucket Bound Hostname HTTP` and `'POST Policy Simple Bucket Bound Hostname'` test data:
```
assert policy["url"] == test_data["policyOutput"]["url"]
E       AssertionError: assert 'http://mydomain.tld' == 'http://mydomain.tld/'
E         - http://mydomain.tld/
E         ?                    -
E         + http://mydomain.tld
```

Reply:
I would consolidate the two cases and prefer conformance test format over api_access_endpoint. 

I don't want to have two paths if possible.